File: src/signaling.py
from types import FunctionType
from typing import Coroutine, Dict
import socketio
from dataclasses import dataclass, asdict

ANSWER_TOPIC = 'answer'
OFFER_TOPIC = 'new_offer'

@dataclass
class OfferOrAnswer:
    uid: int
    d_uid: int
    sdp: str
    con_type: str

@dataclass
class IceCandidate:
    uid: int
    d_uid: int
    candidate: str
    con_type: str

class WebsocketSignalingBase:
    def __init__(self, server_address: str, namespace: str):
        self.server_address = server_address
        self.namespace = namespace
        self.sio = socketio.AsyncClient(ssl_verify=False)

    def add_connections_state_change_handlers(self, on_connect, on_connect_error,
                                              on_disconnect):
        self.sio.on('connect', on_connect, self.namespace)
        self.sio.on('connect_error', on_connect_error, self.namespace)
        self.sio.on('disconnect', on_disconnect, self.namespace)

    def add_handler(self, topic:str, on_message):
        self.sio.on(topic, on_message, self.namespace)

    async def connect_and_wait(self):
        await self.sio.connect(f"{self.server_address}",
                    namespaces=[self.namespace], transports=["websocket"])
        await self.sio.wait()

    async def send(self, topic: str, data):
        await self.sio.emit(topic, data, namespace=self.namespace)


class WebsocketSignaling(WebsocketSignalingBase):
    def __init__(self, server_address: str, namespace: str):
        super().__init__(server_address, namespace)

    async def send_answer(self, uid: int, duid: int, sdp: str, con_type: str):
        my_answer = OfferOrAnswer(uid, duid, sdp, con_type)
        await self.send(ANSWER_TOPIC, asdict(my_answer))


    def add_new_offer_handler(self, on_new_offer):
        async def offer_handler(offer):
            await on_new_offer(int(offer['uid']),
                               int(offer['d_uid']),
                               offer['sdp'],
                               offer['con_type'])

        self.add_handler(OFFER_TOPIC, offer_handler)

File: src/signaling.js
let socket = null;

const SIGNALING_SERVER_ADDRESS = "dinkedpawn.com:4311";
const SIGNALING_SERVER_SOCKETIO_NAMESPACE = "live_cam";

function readUID() {
  return 1;
}

class OfferOrAnswer {
  uid;
  d_uid;
  sdp;
  con_type;

  constructor(uid, d_uid, sdp, con_type) {
    this.uid = uid;
    this.d_uid = d_uid;
    this.sdp = sdp;
    this.con_type = con_type;
  }
}

class IceCandidate {
  uid;
  d_uid;
  candidate;
  con_type;
  constructor(uid, d_uid, candidate, con_type) {
    this.uid = uid;
    this.d_uid = d_uid;
    this.candidate = candidate;
    this.con_type = con_type;
  }
}

class Footage {
  face;
  media;
  constructor(face, media) {
    this.face = face;
    this.media = media;
  }
}

class Face {
  uid;
  faceEncoding;
  isKnown;
  name;
  constructor(uid, faceEncoding, isKnown, name = "") {
    this.uid = uid;
    this.faceEncoding = faceEncoding;
    this.isKnown = isKnown;
    if (name) {
      this.name = name;
    } else {
      this.name = "Unknown";
    }
  }
}

class FootageMedia {
  arrayBuffer;
  mimeType;
  constructor(arrayBuffer, mimeType) {
    this.arrayBuffer = arrayBuffer;
    this.mimeType = mimeType;
  }
}

// function getSignalingServerAddress() {
//   return "dinkedpawn.com:4311";
// }

class SignalingBase {
  sio;
  serverAddress;
  namespace;

  constructor(server_address, namespace) {
    this.serverAddress = server_address;
    this.namespace = namespace;
    this.sio = io(`${this.serverAddress}/${this.namespace}`, {
      transports: ["websocket"],
      autoConnect: false,
    });
  }

  addConnectionStateChangeHandlers(onConnect, onConnectErrror, onDisconnect) {
    this.sio.on("connect", onConnect);
    this.sio.on("connect_error", onConnectErrror);
    this.sio.on("disconnect", onDisconnect);
  }

  addTopicHandler(topic, onMessage) {
    this.sio.on(topic, onMessage);
  }

  connect() {
    this.sio.connect();
  }

  disconnect() {
    this.sio.disconnect();
  }

  // Note the message should be serialized before sending it or it gets
  // deserialized as a json
  send(topic, message) {
    this.sio.emit(topic, message);
  }
}

class Signaling extends SignalingBase {
  constructor(
    serverAddress = SIGNALING_SERVER_ADDRESS,
    namespace = SIGNALING_SERVER_SOCKETIO_NAMESPACE
  ) {
    super(serverAddress, namespace);
  }

  // For live_cam
  addNewAnswerHandler(onAnswerReceived) {
    const newAnswerHandler = async (answerJson) => {
      const answer = new OfferOrAnswer(
        answerJson["uid"],
        answerJson["d_uid"],
        answerJson["sdp"],
        answerJson["con_type"]
      );

      if (answer.d_uid !== readUID()) {
        console.log("Answer is not for this device");
      }

      await onAnswerReceived(answer.uid, answer.sdp, answer.con_type);
    };
    this.addTopicHandler("new_answer", newAnswerHandler);
  }

  sendOffer(d_uid, sdp, con_type) {
    let offer = new OfferOrAnswer(readUID(), d_uid, sdp, con_type);
    this.send("offer", offer);
  }

  // For sent_footage

  addSentFootageHandler(onFootageReceived) {
    const newFootageHandler = async (footageJson) => {
      console.log(footageJson);

      const face = new Face(
        footageJson["uid"],
        footageJson["face_encoding"],
        footageJson["isKnown"],
        footageJson["name"]
      );
        

      const rawFootageData = convertB64ToUint8Array(footageJson['raw_bytes'])

      const media = new FootageMedia(
        rawFootageData,
        footageJson["mimeType"]
      );

      const footage = new Footage(face, media);

      console.log(footage);

      onFootageReceived(footage);
    };
    this.addTopicHandler("new_face_detected", newFootageHandler);
  }

  switchMode(mode) {
    this.send("switch_mode", mode);
  }
}

function convertB64ToUint8Array(b64String){
      const rawFootageData = Uint8Array.from(
        atob(b64String),
        (c) => c.charCodeAt(0)
      );
    return rawFootageData
}

export { Signaling };
